### PR TITLE
Correct Facing Direction for chests & stairs & slabs

### LINKED ISSUE #119: Block placement (stairs, etc.)
BlockTypes representing stairs, slabs, logs, leaves, ladders, and other assorted blocks with facing need to be created to allow these to be placed correctly. The `placeBlock` function should be overridden to modify the passed BlockState to set data based on the player's facing or the face placed against as needed.

Blocks involved:
- [x] Torches
- [ ] Redstone torches
- [x] Logs
- [ ] Leaves
- [x] Slabs
- [ ] Tall flowers
- [ ] Pistons
- [x] Stairs
- [X] Sign posts and wall signs
- [x] Doors
- [ ] Rails
- [x] Ladders
- [x] Furnaces
- [x] Chests / Ender chests
- [x] Dispensers
- [x] Droppers
- [x] Hoppers
- [x] Levers
- [x] Buttons
- [ ] Pumpkins / Jack o'Lanterns
- [ ] Redstone repeaters / comparators
- [x] Trap doors
- [x] Vines
- [x] Fence gates
- [ ] Cocoa beans
- [ ] Tripwire hooks
- [x] Heads
- [x] Hay
- [x] Quartz
- [x] Beds
- [ ] Anvils
- [x] Banners

Edit (turt2live): Issue scope expanded to general placement problems including data not being set (leaves, etc) and not just facing directions.

### PR DESCRIPTION
**Pull request by ZachBora**
The getOppositeDirection function could probably be put elsewhere to be
reused. This one is only for 4 faces.

Partially fixes #119

### CODE REVIEW

**Comment by SpaceManiac** [comment]
Looks solid. I will test and merge along with the other blocktype additions.

**Comment by ZachBora** [comment]
Added slabs and enderchest

File: src/main/java/net/glowstone/block/ItemTable.java [hunk 1]
```diff
@@ -84,7 +84,21 @@ private void registerBuiltins() {
...
+        reg(Material.JUNGLE_WOOD_STAIRS, new BlockStairs());
+        reg(Material.NETHER_BRICK_STAIRS, new BlockStairs());
+        reg(Material.QUARTZ_STAIRS, new BlockStairs());
+        reg(Material.SANDSTONE_STAIRS, new BlockStairs());
+        reg(Material.SPRUCE_WOOD_STAIRS, new BlockStairs());
+        reg(Material.SMOOTH_STAIRS, new BlockStairs());
+        reg(Material.WOOD_STAIRS, new BlockStairs());
+        reg(Material.STEP, new BlockSlab());
+        reg(Material.WOOD_STEP, new BlockSlab());
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/ItemTable.java]: Whitespace

File: src/main/java/net/glowstone/block/blocktype/BlockChest.java [hunk 2]
```diff
@@ -25,12 +25,33 @@ public void placeBlock(GlowPlayer player, GlowBlockState state, BlockFace face,
...
             state.setData(data);
         } else {
             // complain?
             GlowServer.logger.warning("Placing Chest: MaterialData was of wrong type");
         }
     }
+    
+    public BlockFace getOppositeDirection(GlowPlayer player) {
+        double rot = player.getLocation().getYaw() % 360;
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockChest.java]: Whitespace

File: src/main/java/net/glowstone/block/blocktype/BlockChest.java [hunk 3]
```diff
@@ -25,12 +25,33 @@ public void placeBlock(GlowPlayer player, GlowBlockState state, BlockFace face,
...
         }
     }
+    
+    public BlockFace getOppositeDirection(GlowPlayer player) {
+        double rot = player.getLocation().getYaw() % 360;
+        
+        if (rot < 0) {
+            rot += 360.0;
+        }
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockChest.java]: Whitespace

File: src/main/java/net/glowstone/block/blocktype/BlockChest.java [hunk 4]
```diff
@@ -25,12 +25,33 @@ public void placeBlock(GlowPlayer player, GlowBlockState state, BlockFace face,
...
-            // todo: determine facing direction
-            ((Chest) data).setFacingDirection(BlockFace.EAST);
+            ((Chest) data).setFacingDirection(getOppositeDirection(player));
             state.setData(data);
         } else {
             // complain?
             GlowServer.logger.warning("Placing Chest: MaterialData was of wrong type");
         }
     }
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockChest.java]: Whitespace

File: src/main/java/net/glowstone/block/blocktype/BlockChest.java [hunk 5]
```diff
@@ -25,12 +25,33 @@ public void placeBlock(GlowPlayer player, GlowBlockState state, BlockFace face,
...
-            ((Chest) data).setFacingDirection(BlockFace.EAST);
+            ((Chest) data).setFacingDirection(getOppositeDirection(player));
             state.setData(data);
         } else {
             // complain?
             GlowServer.logger.warning("Placing Chest: MaterialData was of wrong type");
         }
     }
+    
+    public BlockFace getOppositeDirection(GlowPlayer player) {
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockChest.java]: There are several PRs which handle "looking backwards". Maybe it's ideal to move this to a util call?

File: src/main/java/net/glowstone/block/blocktype/BlockEnderchest.java [hunk 6]
```diff
@@ -29,12 +29,32 @@ public void placeBlock(GlowPlayer player, GlowBlockState state, BlockFace face,
...
-            // todo: determine facing direction
-            ((EnderChest) data).setFacingDirection(BlockFace.EAST);
+            ((EnderChest)data).setFacingDirection(getOppositeDirection(player));
             state.setData(data);
         } else {
             // complain?
             GlowServer.logger.warning("Placing EnderChest: MaterialData was of wrong type");
         }
     }
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockEnderchest.java]: Whitespace

File: src/main/java/net/glowstone/block/blocktype/BlockSlab.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+import net.glowstone.entity.GlowPlayer;
+import org.bukkit.Material;
+import org.bukkit.block.BlockFace;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.material.MaterialData;
+import org.bukkit.material.Step;
+import org.bukkit.material.WoodenStep;
+import org.bukkit.util.Vector;
+
+public class BlockSlab extends BlockType {
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockSlab.java:14]: For consistency, there should be a blank line after this declaration.

File: src/main/java/net/glowstone/block/blocktype/BlockStairs.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+import net.glowstone.GlowServer;
+import net.glowstone.block.GlowBlockState;
+import net.glowstone.entity.GlowPlayer;
+import org.bukkit.block.BlockFace;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.material.MaterialData;
+import org.bukkit.material.Stairs;
+import org.bukkit.util.Vector;
+
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockStairs.java]: Whitespace

File: src/main/java/net/glowstone/block/blocktype/BlockStairs.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+public class BlockStairs extends BlockType {
+
+    @Override
+    public void placeBlock(GlowPlayer player, GlowBlockState state, BlockFace face, ItemStack holding, Vector clickedLoc) {
+        super.placeBlock(player, state, face, holding, clickedLoc);
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockStairs.java]: Whitespace

File: src/main/java/net/glowstone/block/blocktype/BlockStairs.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+public class BlockStairs extends BlockType {
+
+    @Override
+    public void placeBlock(GlowPlayer player, GlowBlockState state, BlockFace face, ItemStack holding, Vector clickedLoc) {
+        super.placeBlock(player, state, face, holding, clickedLoc);
+        
+        MaterialData data = state.getData();
+        if (data instanceof Stairs) {
+            ((Stairs) data).setFacingDirection(getOppositeDirection(player));
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockStairs.java]: Whitespace

File: src/main/java/net/glowstone/block/blocktype/BlockStairs.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+            ((Stairs) data).setFacingDirection(getOppositeDirection(player));
+            
+            if (face == BlockFace.DOWN || face != BlockFace.UP && clickedLoc.getY() >= 8) {
+                ((Stairs) data).setInverted(true);   
+            }
+
+            state.setData(data);
+        } else {
+            // complain?
+            GlowServer.logger.warning("Placing Stairs: MaterialData was of wrong type");
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockStairs.java:29]: This is inconsistent with other places

File: src/main/java/net/glowstone/block/blocktype/BlockEnderchest.java [hunk 12]
```diff
@@ -29,12 +29,32 @@ public void placeBlock(GlowPlayer player, GlowBlockState state, BlockFace face,
...
-            ((EnderChest) data).setFacingDirection(BlockFace.EAST);
+            ((EnderChest)data).setFacingDirection(getOppositeDirection(player));
             state.setData(data);
         } else {
             // complain?
             GlowServer.logger.warning("Placing EnderChest: MaterialData was of wrong type");
         }
     }
+    
+    public BlockFace getOppositeDirection(GlowPlayer player) {
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockEnderchest.java]: https://github.com/SpaceManiac/Glowstone/pull/166/files#r17219136

File: src/main/java/net/glowstone/block/blocktype/BlockStairs.java (hunk 11 above)
Comment (ZachBora) [inline src/main/java/net/glowstone/block/blocktype/BlockStairs.java:29]: That's how stairs work, not sure how you'd do it differently.

Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockStairs.java:29]: I mean the message is inconsistent with the rest of the project.

Looks like it should be something like
`GlowServer.logger.warning("Interacting " + getMaterial().name() + ": MaterialData was of wrong type (" + data.getClass().getName() + ")");`

Comment (ZachBora) [inline src/main/java/net/glowstone/block/blocktype/BlockStairs.java:29]: It's the stair class and I did the warning just like in the other block classes.

**Comment by SpaceManiac** [comment]
Chests and stairs look pretty solid. I'm not sure about the slab handling right now, though.

Would it make sense for BlockType to have a general mechanism to allow blocks to absorb the placement of other blocks on them? (general handling for tall grass as well as slabs). I think this would be a more elegant design than the current special-casing for slabs.

File: src/main/java/net/glowstone/block/blocktype/BlockChest.java (hunk 5 above)
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockChest.java]: https://github.com/SpaceManiac/Glowstone/blob/master/src/main/java/net/glowstone/block/blocktype/BlockType.java#L172

File: src/main/java/net/glowstone/block/blocktype/BlockEnderchest.java (hunk 12 above)
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockEnderchest.java]: https://github.com/SpaceManiac/Glowstone/blob/master/src/main/java/net/glowstone/block/blocktype/BlockType.java#L172

File: src/main/java/net/glowstone/block/blocktype/BlockSlab.java [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+                return SlabPlaceable.AGAINSTBLOCK;
+            }
+            if ((target.getType() == holding.getType()) && (target.getData() % 8 == holdingData)) {
+                return SlabPlaceable.TARGETBLOCK;
+            }
+        }
+        return SlabPlaceable.NO;
+    }
+
+    public static enum SlabPlaceable {
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockSlab.java]: Oracle prefers enums at the top of classes.

File: src/main/java/net/glowstone/block/blocktype/BlockStairs.java (hunk 11 above)
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockStairs.java:29]: > .< Fine.

File: src/main/java/net/glowstone/block/blocktype/BlockStairs.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+        }
+    }
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockStairs.java]: Whitespace

File: src/main/java/net/glowstone/block/blocktype/BlockStairs.java [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+            state.setData(data);
+        } else {
+            // complain?
+            GlowServer.logger.warning("Placing Stairs: MaterialData was of wrong type");
+        }
+    }
+    
+    public BlockFace getOppositeDirection(GlowPlayer player) {
+        double rot = player.getLocation().getYaw() % 360;
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockStairs.java]: Whitespace

File: src/main/java/net/glowstone/block/blocktype/BlockStairs.java [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+        }
+    }
+    
+    public BlockFace getOppositeDirection(GlowPlayer player) {
+        double rot = player.getLocation().getYaw() % 360;
+        
+        if (rot < 0) {
+            rot += 360.0;
+        }
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockStairs.java]: Whitespace

File: src/main/java/net/glowstone/block/blocktype/BlockStairs.java [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+            }
+
+            state.setData(data);
+        } else {
+            // complain?
+            GlowServer.logger.warning("Placing Stairs: MaterialData was of wrong type");
+        }
+    }
+    
+    public BlockFace getOppositeDirection(GlowPlayer player) {
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockStairs.java]: https://github.com/SpaceManiac/Glowstone/blob/master/src/main/java/net/glowstone/block/blocktype/BlockType.java#L172

File: src/main/java/net/glowstone/block/blocktype/BlockType.java [hunk 18]
```diff
@@ -116,11 +116,13 @@ public boolean blockInteract(GlowPlayer player, GlowBlock block, BlockFace face,
     public final void rightClickBlock(GlowPlayer player, GlowBlock against, BlockFace face, ItemStack holding, Vector clickedLoc) {
         GlowBlock target = against.getRelative(face);
         GlowBlockState newState = target.getState();
+        BlockSlab.SlabPlaceable slabplaceable = BlockSlab.isSlabPlaceable(player, target, against, face, holding);
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockType.java]: Incorrect capitalization on the variable name

File: src/main/java/net/glowstone/block/blocktype/BlockSlab.java [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+        NO, TARGETBLOCK, AGAINSTBLOCK;
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockSlab.java]: `NO` - Needs a more descriptive name
`TARGETBLOCK` and `AGAINSTBLOCK` need to have underscores

**Comment by SpaceManiac** [comment]
Merged in 61914c1.

**Pull request closed without merging**